README
Modwheel-Apache is the Apache2+mod_perl bindings for Modwheel.

See the file config/modwheelapache.conf for a sample configuration file for running
Modwheel with apache2.

* Requirements
    Apache2                            - http://httpd.apache.org
    mod_perl2                        - http://perl.apache.org
    libapreq2                        - http://search.cpan.org/~joesuf/libapreq2-2.08/

* Installation

Module::Install uses the standard perl module install process:

    cpansign -v         # optional; see SIGNATURE for details
    perl Makefile.PL
    make                # or 'nmake' on Win32
    make test
    make install

* Notes for Windows users

If you are using Microsoft Windows, you will probably need the 'nmake'
utility before installation. It's available at:

    ftp://ftp.microsoft.com/Softlib/MSLFILES/nmake15.exe

If you are connected to the internet, "perl Makefile.PL" is capable
to automatically fetch it, and install into your windows directory;
all you have to do is to answer 'y' to this question:

    Required executable 'nmake' not found. Install it? [Y/n]:

However, if you are not yet online, or wish to install 'nmake' into
another directory, then please download the file manually, save it to
a directory in %PATH (e.g. C:\WINDOWS\COMMAND), then launch the MS-DOS
command line shell, "cd" to that directory, and run "nmake15.exe" from
there; that will create the 'nmake.exe' file needed by CPANPLUS.

You may then resume the installation process above, but replace 'make'
with 'nmake' in the last three lines.

* Copyright, contact and licensing information.

Copyright 2007 by
    Ask Solem Hoel <ask@0x61736b.net>

All rights reserved.  You can redistribute and/or modify
this bundle under the same terms as Perl itself.

See <http://www.perl.com/perl/misc/Artistic.html>.


Changes
0.2.3 Sat May 19 00:55:12 [ASKSH@cpan.org]
    - now compatible with Modwheel 0.2.3 (and 0.3.1)
    - small bug fixes.
0.1 Mon Apr 2 19:02:00  Ask Solem Hoel
    - Initial Version


Build.PL
use strict;
use warnings;
use inc::Module::Build::Modwheel;

sub YES { 1 };
sub NO  { 0 };

# if set to a string, we should create the installation directory.
my $prefix = q{};

my $builder = Module::Build->new(
    module_name          => 'Apache2-Modwheel',
    license              => 'perl',
    dist_author          => 'Ask Solem <ASKSH@cpan.org>',
    dist_version_from    => 'lib/Apache2/Modwheel.pm',
    dynamic_config       => NO,
    sign                 => NO,
    recursive_test_files => YES,
    requires             => {
        'Modwheel'              => '0.2.3',
        'mod_perl2'             => 2.00,
        'Tie::IxHash'           => 1.21,
        'Parse::RecDescent'     => 1.94,
        'ExtUtils::XSBuilder::ParseSource' => 0.03,
        'Apache2::Request'      => 2.00,
        'Apache2::Upload'       => 2.00,
        'namespace::clean'      => 0,
        'version'               => 0,
        'Carp'                  => 0,
        'Scalar::Util'          => 0,
        'Readonly'              => 1.00,
    },
    recommends           => {
        'Test::Pod'             => 0,
        'Pod::Coverage'         => 0,
        'Test::Pod::Coverage'   => 0,
        'Readonly::XS'          => 1.00,
        'Test::Kwalitee'        => 0.30,
        'Test::YAML::Meta'      => 0.004,
    },
    build_requires       => {
        'Test::More'            => 0.42,
    },
    add_to_cleanup       => [ qw(
        a.out
        test.pl
        test.c
        test.cpp
        test.m
        *.swp
        .gdb_history
        install.cache
        t/cache
    ) ],
    meta_merge          => {
        resources           => {
            HomePage            => 'http://www.0x61736b.net/Modwheel',
            Download            => 'http://search.cpan.org/~asksh/Apache2-Modwheel-0.2.3/',
        },
        distribution_type   => 'Application',
    },
    
);

$builder->create_build_script( );



devel/OLD.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package Apache2::Modwheel;
use strict;
use warnings;
use 5.00800;
use mod_perl2 2.0;
use version; our $VERSION = qv('0.0.2');

use Readonly;
use Apache2::Access            ();
use Apache2::RequestRec     ();
use Apache2::RequestUtil     ();
use Apache2::RequestIO        ();
use Apache2::Request        ();
use Apache2::Upload            ();
use Apache2::Const            -compile =>
    qw(OK SERVER_ERROR FORBIDDEN NOT_FOUND DECLINED);
use Modwheel                ();
use Modwheel::DB             ();
use Modwheel::User            ();
use Modwheel::Object        ();
use Modwheel::Template        ();
use Modwheel::Repository    ();
use Params::Util ('_HASH', '_ARRAY', '_CODELIKE', '_INSTANCE');

# 10MB's Upload Limit.
Readonly my $MAX_UPLOAD_LIMIT => 10_485_760;

sub handler : method {
    my ($class, $r) = @_;
    return Apache2::Const::DECLINED if dont_want_this($r);

    # Don't process header requests.
    return Apache2::Const::OK        if $r->header_only();

    my $disable_uploads =
        $r->dir_config('ModwheelFileUploads') =~ m/yes/xmsi
        ? 0
        : 1;
    my $req = Apache2::Request->new($r,
        #POST_MAX => "10M",
        MAX_BODY => $MAX_UPLOAD_LIMIT,
        DISABLE_UPLOADS => $disable_uploads,
    );

    # Create log handler routine for Modwheel::log_*.
    my $loghandler_apache2 = sub  {
        my ($modwheel, $log_string, $facility) = @_;
        my $apache2 = $modwheel->logobject;
        my $rlog = $apache2->log;

        #*__ANON__ = 'Apache Modwheel Loghandler()';
        if ($facility eq 'Error') {
            return $rlog->error($log_string);
        }
        if ($facility eq 'Warning') {
            return $rlog->warn($log_string);
        }
        if ($facility eq 'Info') {
            return $rlog->info($log_string);
        }
        if ($facility eq 'Debug') {
            return $rlog->debug($log_string);
        }

        return;
    };

    # Set up our Modwheel instance.

    my $modwheel_config = {
        prefix             => $r->dir_config('ModwheelPrefix'),
        configfile         => $r->dir_config('ModwheelConfigFile'),
        site             => $r->dir_config('ModwheelSite'),
        locale             => $r->dir_config('Locale'),
        add_loghandlers  => {apache2         => $loghandler_apache2,},
        logobject        => $r,
        logmode             => 'apache2',
        logmode          => 'stderr',
    };

    my $modwheel = Modwheel->new($modwheel_config);
    my $db       = Modwheel::DB->new({modwheel => $modwheel,});
    my $user     = Modwheel::User->new({
        modwheel => $modwheel,
        db         => $db,
    });
    my $object     = Modwheel::Object->new({
       modwheel => $modwheel,
       db          => $db,
       user      => $user,
    });
    my $repository = Modwheel::Repository->new({
        modwheel => $modwheel,
        db         => $db,
        user     => $user,
    });
    my $template = Modwheel::Template->new({
        modwheel => $modwheel,
        db         => $db,
        user     => $user,
        object   => $object,
        repository => $repository,
    });

    # Connect to database
    my $no_connect = $r->dir_config('NoDatabaseConnect');
    if ($no_connect ne 'Yes') {
        $db->connect() or return 500; #Apache2::Const::SERVER_ERROR
    }

    # Save user info if user is logged in.
    my $uname   = $r->user;
    $uname ||= 'guest';
    $user->set_uname($uname);
    if ($db->connected) {
        $user->set_uid( $user->uidbyname($uname) );
    }

    # Set up parent id or convert web path to parent id.
    my $parent;
    my $page = $r->uri;

    # untaint location.
    my $loc  = quotemeta $r->location;
    my $useWebPath = $r->dir_config('ModwheelWebPathToId');
    if ($useWebPath && $useWebPath =~ /yes/xmsi && $db->connected) {
        $parent = $req->param('parent');
        if ($parent) {

            # remove location part of uri requested.
            $page =~ s/^ $loc \/?//xms;
        }
        else {
            $parent = $object->path_to_id($page);

            unless ($parent) {
                $db->disconnect();
                return Apache2::Const::NOT_FOUND;
            }
        }
        $page =~ s{ ^.*/ }{}xms;
        undef $page unless $page =~ m/ \.[\w\d_]+$ /xms;
    }
    else {
        $parent =  $req->param('parent');

        # remove location part of uri requested.
        $page   =~ s/^ $loc \/?//xms;
    }
    $parent ||= Modwheel::Object::MW_TREE_ROOT;

    # Find the filename (page) for this object
    if ($r->dir_config('ModwheelFollowTemplates')) {
        my $o = $object->fetch({id => $parent});
        $page = $o->template if $o->template;
    }
    $page   ||= $modwheel->siteconfig->{directoryindex};

    # remove leading slash.
    $page =~ s{^/}{}xms;

    # add template dir as prefix to page path
    $page = $modwheel->siteconfig->{templatedir} . q{/} . $page;
    return Apache2::Const::NOT_FOUND unless -f $page;

    if (!$disable_uploads && $db->connected) {
        handle_uploads($r, $req, $db, $repository, $parent);
    }

    # set up the template object:
    $template->init(
        {   input  => $page,
            param  => $req,
            parent => $parent,
        }
    ) or return printError($template->errstr);

    my $content_type = $r->dir_config('ContentType') || 'text/html';
    $r->content_type($content_type);

    my $process_args = {};
    my $output = $template->process($process_args);

    $r->print($output);

    $db->disconnect() if $db->connected;

    return Apache2::Const::OK;

}

sub dont_want_this {
    my ($r) = @_;

    # we don't care about favicon.ico XXX: this is a hack, must change
    return 1 if $r->uri =~ m/ favicon.ico$ /xms;

    my $dontHandle = $r->dir_config('DontHandle');
    if ($dontHandle) {
        my @dont_handle_locations = split m/ \s+ /xms, $dontHandle;
        foreach my $location (@dont_handle_locations) {

            # remove trailing slashes.
            $location =~ s{ ^/ }{}xms;
            return 1 if $r->uri =~ m{ ^/ \Q$location\E }xms;
        }
    }

    return 0;
}
   
sub handle_uploads( ) {
    my ($r, $req, $db, $repository, $parent) = @_;

    # If the user uploads files, add them to the repository.
    my @uploads;
    foreach ($req->upload) {
        my $upload = $req->upload($_);
        if ($upload) {
            my $upload_in = $req->param('id');
            $upload_in ||= $parent;
            my %current_upload = (
                filename => $upload->filename,
                mimetype => $upload->type,
                size     => $upload->size,
                parent     => $upload_in
            );
            $repository->upload_file($upload->fh, %current_upload);
            push @uploads, \%current_upload;
        }
    }

    return;
}

sub printError {
    my ($r, $errstr) = @_;
    $r->content_type('text/html');
    print <<"HTML"
    <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
    <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
        <head>
            <title>Modwheel - Error</title>
            <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
        </head>
    <body>
    <h1>Sorry! An error occured.</h1>
    <h3>The error was:</h3>
    <p>$errstr</p>
    </body>
    </html>
HTML
        ;

    return Apache2::Const::OK;
}

1;
__END__

=head1 NAME

Apache2::Modwheel - Use Modwheel with mod_perl2

=head1 SYNOPSIS

    <VirtualHost *:80>
            ServerName admin.localhost
        ErrorLog logs/error_log
        <Location />
            SetHandler perl-script                                                                                                             
            PerlAuthenHandler   Apache2::Modwheel::Authen                                                                                     
            PerlResponseHandler Apache2::Modwheel                                                                                             
            PerlSetVar ModwheelPrefix       /opt/devel/Modwheel                                                                               
            PerlSetVar ModwheelConfigFile   config/modwheelconfig.yml                                                                         
            PerlSetVar ModwheelSite         Admin                                                                                             
            PerlSetVar ModwheelFileUploads  Yes                                                                                               
            PerlSetVar Locale               en_EN                                                                                             
            PerlSetVar DontHandle           "rep javascript css images scriptaculous"                                                         
                                                                                                                                          
            AuthType Basic                                                                                                                    
            AuthName "void"                                                                                                                   
            Require valid-user                                                                                                                
        </Location>                                                                                                                           
        Alias /rep /opt/devel/Modwheel/Repository                                                                                             
        Alias /css /opt/devel/Modwheel/Templates/SimpleAdmin/css                                                                              
        Alias /javascript /opt/devel/Modwheel/Templates/SimpleAdmin/javascript                                                                
        Alias /scriptaculous /opt/devel/Modwheel/Templates/Scriptaculous                                                                      
        <Directory /opt/devel/Modwheel/Repository/*/*>                                                                                        
            Order Deny,Allow                                                                                                                  
            Allow from all                                                                                                                    
        </Directory>                                                                                                                          
        <Directory /opt/devel/Modwheel/Templates/*/*>                                                                                         
            Order Deny,Allow                                                                                                                  
            Allow from all                                                                                                                    
        </Directory>                                                                                                                          
    </VirtualHost>    


=cut

# Local variables:
# vim: ts=4

               M O D W H E E L for A P A C H E 2

INSTALLING THE Modwheel-Apache2 BINDINGS

Installation of the modules is easy:

    perl Makefile.PL
    make
    make install

After this you have to configure your apache server.
The Modwheel install script should have created an example
httpd.conf example to your Modwheel installation dir.
The file is named config/Modwheel-apache-example.conf and can also
be found in the config directory of this distribution.
Study this file and change it to fit your configuration, then add it to your
httpd.conf. Here is a list of what the new Apache configuration directives
means:


=head1 HANDLERS

=over 4

=item PerlResponseHandler L<Apache2::Modwheel>

This is the main Modwheel handler. It requires that you have
the ModwheelPrefix and ModwheelConfigFile options set.

=item PerlAuthenHandler L<Apache2::Modwheel::Authen>

This module is for authentication via the modwheel user system.

=back

=head1 APACHE CONFIGURATION DIRECTIVES

=over 4

=item PerlSetVar C<ModwheelPrefix>

This is the directory you installed Modwheel to.

Example:

    PerlSetVar ModwheelPrefix "/opt/Modwheel"

=item PerlSetVar C<ModwheelConfigFile>

This is the Modwheel configuration file to use.
If the filename is relative Modwheel will search for it in
the ModwheelPrefix directory.

Example:

    PerlSetVar ModwheelConfigFile "config/Modwheelconfig.yml"

=item PerlSetVar C<ModwheelSite>

The site to use. Sites are configured in the configuration file.

Example:

    PerlSetVar ModwheelSite "SimpleAdmin"

=item PerlSetVar C<ModwheelFileUploads>

Users will be able to upload files to the site if this is set to Yes.
This is used for the Repositories.

Example:

    PerlSetVar ModwheelFileUploads Yes

=item PerlSetVar C<ModwheelWebPathToId>

If this is set to yes, a user can enter i.e
        http://foo.bar/Music/Aphex Twin
in his browser and Modwheel will find the node in the object tree
with this name.

Example:

    PerlSetVar ModwheelWebPathToId Yes

=item PerlSetVar C<Locale>

What language the site is in. For a list of the values that are possible
with this directive you can enter this command: (if you are running a form
of Unix):

    locale -a

Example:

    PerlSetVar Locale en_EN

=item PerlSetVar C<DontHandle>

URL's that Modwheel should'nt handle for this site. This is meant for
static content that does not need any content. i.e /images, /javascript
and so on.

Example:

    PerlSetVar DontHandle "rep javascript css images scriptaculous"

=item PerlSetVar C<NoDatabaseConnect>

Modwheel will not connect to the database if this is set to 'Yes'.

Example:

    PerlSetVar NoDatabaseConnect Yes

=item PerlSetVar C<ContentType>

Set the content type for this site.

Example:

    PerlSetVar ContentType "text/html"

=item PerlSetVar C<ModwheelFollowTemplates>

If a object has a user-specified template defined and ModwheelFollowTemplates
is set to 'Yes', it will choose this template instead of the default.

Example:

    PerlSetVar ModwheelFollowTemplates Yes

=back

=head1 EXPORT

None.

=head1 HISTORY

=over 8

=item 0.01

Initial version.

=back

=head1 SEE ALSO

=over 4

=item * L<Modwheel::Manual::Install>

=item * L<Modwheel::Manual::Config>

=item * L<Modwheel::Manual::Intro>

=item * L<Apache2::Modwheel::Authen>

=item * The README included in the Modwheel distribution.

=item * The Modwheel website: L<http://www.0x61736b.net/Modwheel/>

=back

=head1 AUTHORS

Ask Solem, F<< ask@0x61736b.net >>.

=head1 COPYRIGHT, LICENSE

Copyright (C) 2007 by Ask Solem Hoel C<< ask@0x61736b.net >>.

All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.



BUILD-ubuntu.sh
#!/bin/bash

echo "--- Install Modwheel-Apache2"
echo "This will install Apache2 with mod_perl and libapreq enabled"
echo "in /opt/apache2, it requires that you already installed Modwheel in the"
echo "location /opt/modwheel."
echo 
echo "[Press ENTER to continue]"
read

build_dir="./build-modwheel-apache2-$$"
mkdir "$build_dir"
cd "$build_dir"

# ### Install Apache 2.2.4
(
wget http://www.powertech.no/apache/dist/httpd/httpd-2.2.4.tar.gz
tar xvfz httpd-2.2.4.tar.gz
cd httpd-2.2.4
./configure --prefix=/opt/apache2 --with-mpm=prefork
make
sudo make install
)

# ### Install mod_perl 2.0.3
(
wget http://perl.apache.org/dist/mod_perl-2.0-current.tar.gz
tar xvfz mod_perl-2.0-current.tar.gz
cd mod_perl-2.0.3
perl Makefile.PL MP_APXS=/opt/apache2/bin/apxs
make && sudo make install
)

# ### Install Tie::IxHash 1.21 (required by libapreq)
(
wget http://search.cpan.org/CPAN/authors/id/G/GS/GSAR/Tie-IxHash-1.21.tar.gz
tar xvfz Tie-IxHash-1.21.tar.gz
cd Tie-IxHash-1.21
perl Makefile.PL && sudo make install
)

# ### Install Parse::RecDescent 1.94 (required by libapreq)
(
wget http://search.cpan.org/CPAN/authors/id/D/DC/DCONWAY/Parse-RecDescent-1.94.tar.gz
tar xvfz Parse-RecDescent-1.94.tar.gz
cd Parse-RecDescent-1.94
perl Makefile.PL && sudo make install
)

# ### Install ExtUtils::XSBuilder 0.28 (required by libapreq)
(
wget http://search.cpan.org/CPAN/authors/id/G/GR/GRICHTER/ExtUtils-XSBuilder-0.28.tar.gz
tar xvfz ExtUtils-XSBuilder-0.28.tar.gz
cd ExtUtils-XSBuilder-0.28
perl Makefile.PL && sudo make install
)

# ### Install the expat XML-Parser 2.0.0 (required by libapreq)
(
wget http://ovh.dl.sourceforge.net/sourceforge/expat/expat-2.0.0.tar.gz
tar xvfz expat-2.0.0.tar.gz
cd expat-2.0.0
./configure
make && sudo make install
)

# ### Install libconv 1.11 (required by libapreq)
(
wget http://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.11.tar.gz
tar xvfz libiconv-1.11.tar.gz
cd libiconv-1.11
./configure
make && sudo make install
)

# ### Install apreq2
(
wget http://search.cpan.org/CPAN/authors/id/J/JO/JOESUF/libapreq2-2.08.tar.gz
tar xvfz libapreq2-2.08.tar.gz
cd libapreq2-2.08
perl Makefile.PL
make && sudo make install
)

(
wget http://www.0x61736b.net/Modwheel/Modwheel-Apache2-0.01.tar.gz
tar xvfz Modwheel-Apache2-0.01.tar.gz
cd Modwheel-Apache2-0.01
perl Makefile.PL && sudo make install
)

# ### Add LoadModule configuration directives to apache configuration
sudo sh -c 'echo "LoadModule apreq_module /opt/apache2/modules/mod_apreq2.so" >> /opt/apache2/conf/httpd.conf'
sudo sh -c 'echo "LoadModule perl_module  /opt/apache2/modules/mod_perl.so" >> /opt/apache2/conf/httpd.conf'

# ### Add the automaticly generated Modwheel apache config.
sudo sh -c 'cat /opt/modwheel/config/modwheel-apache-example.conf >> /opt/apache2/conf/httpd.conf'

# UBUNTU
sudo killall apache2
sudo chmod 000 /etc/init.d/apache2

sudo /opt/apache2/bin/apachectl start


lib/Apache2/Modwheel/Authen.pm

package Apache2::Modwheel::Authen;
# ===================================================================== #
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
# ===================================================================== #
use strict;
use warnings;
use 5.008;
use mod_perl2 2.0;
use version; our $VERSION = qv('0.2.3');

use Readonly;
use Modwheel                ();
use Modwheel::DB            ();
use Modwheel::User          ();
use Apache2::Access         ();
use Apache2::Connection     ();
use Apache2::RequestUtil    ();
use Apache2::RequestRec     ();
use Apache2::ServerRec      ();
use Apache2::Log            ();
use Apache2::Const          -compile
    => qw(OK AUTH_REQUIRED HTTP_UNAUTHORIZED DECLINED SERVER_ERROR);
use namespace::clean;

# --------------------------------------------------------------------- #
# > loghandler_authen_apache2
#
# Loghandler for Modwheel.
# --------------------------------------------------------------------- #
Readonly my $LOGHANDLER_AUTHEN_APACHE2 => sub {

    my ($modwheel, $log_message) = @_;
    my $r = $modwheel->logobject;

    return $r->log_error($log_message);
};



# --------------------------------------------------------------------- #
# > handler($class, $r);
#
# Log in to a Modwheel system.
# --------------------------------------------------------------------- #
sub handler : method {
    my ($class, $r) = @_;
    return Apache2::Const::DECLINED if !$r;

    my ($res, $sent_pw) = $r->get_basic_auth_pw();
    return $res if $res != Apache2::Const::OK;

    my $uname = $r->user;
    if (!$uname || !$sent_pw) {
        $r->note_basic_auth_failure;
        $r->log_reason('Need both username and password.');
        return Apache2::Const::HTTP_UNAUTHORIZED;
    }

    my $modwheel_config = {
        prefix          => $r->dir_config('ModwheelPrefix'),
        configfile      => $r->dir_config('ModwheelConfigFile'),
        site            => $r->dir_config('ModwheelSite'),
        locale          => $r->dir_config('Locale'),
        add_loghandlers => {a2authen => $LOGHANDLER_AUTHEN_APACHE2},
        logmode         => 'a2authen',
        logobject       => $r,
    };

    my $modwheel = Modwheel->new(
        $modwheel_config,
    );
    my $db       = Modwheel::DB->new({
        modwheel => $modwheel,
    });
    my $user     = Modwheel::User->new({
        modwheel => $modwheel,
        db       => $db,
    });
   
    my $connection_config_ref;
    if ($r->dir_config('ModwheelCachedDB') eq 'Yes') {
        $connection_config_ref->{cached} = 1;
    }
    $db->connect( $connection_config_ref )
        or return Apache2::Const::SERVER_ERROR;
    my $server          = $r->server;
    my $client          = $r->connection;
    my $server_hostname = $server->server_hostname;
    my $auth_type       = $r->ap_auth_type;
    my $site            = $modwheel->site;
    my $remote_addr     = $client->remote_ip;
    if (! $user->login($uname, $sent_pw, $remote_addr)) {
        $r->note_basic_auth_failure;
        $r->log_reason( $modwheel->error );
        $db->disconnect( );
        $r->log_reason(
            "Invalid login to site $site from $remote_addr " .
            "using username: $uname"
        );
        return Apache2::Const::HTTP_UNAUTHORIZED;
    }


    #$r->warn("Modwheel [site: $site@$server_hostname] info || ",
    #    "| Login: $uname [ip: $remote_addr, auth type: $auth_type]"
    #);    

    return Apache2::Const::OK;
}


1;


Makefile.PL
use strict;
use warnings;
no strict 'subs';
use English qw( -no_match_vars );
use ExtUtils::MM_Unix ();

=for comment

BEGIN {
    eval 'use version';
    if ($EVAL_ERROR) {
        die 'Modwheel requires the version module. '            .
            'You can download version from CPAN: '              .
            "http://search.cpan.org/~jpeacock/version-0.71/.\n"
        ;
    }
    eval "use Readonly";
    if ($EVAL_ERROR) {
        die 'Apache2::Modwheel requires the Readonly module. '  .
            'You can download Readonly from CPAN: '             .
            "http://search.cpan.org/~roode/Readonly-1.03/\n"
        ;
    }
}

use version;
use Readonly;

my $v = version->new;

Readonly my $MAIN_DIST_FILE => 'lib/Apache2/Modwheel.pm';

my $modwheel_version
    = ExtUtils::MM_Unix->parse_version($MAIN_DIST_FILE);
my $modwheel_numified_version
    = version->new($modwheel_version)->numify;

=cut

use inc::Module::Install;
    
# Define metadata
name            'Apache2-Modwheel';
version         '0.2.3';
all_from        'lib/Apache2/Modwheel.pm';
author          'Ask Solem <ask@0x61736b.net>';
license         'perl';

    
# Specific dependencies
requires        'namespace::clean'          => 0;
requires        'version'                   => 0;
requires        'Carp'                      => 0;
requires        'Scalar::Util'              => 1.14;
requires        'Readonly'                  => 1.00;
requires        'Modwheel'                  => 0.000002;
requires        'mod_perl2'                 => 2.00;
requires        'Tie::IxHash'               => 1.21;
requires        'Parse::RecDescent'         => 1.94;
requires        'ExtUtils::XSBuilder::ParseSource' => 0.03;
requires        'Apache2::Request'          => 2.00;
requires        'Apache2::Upload'           => 2.00;

build_requires  'Test::More'                => 0.42;
recommends      'Test::YAML::Meta'          => 0.04;
recommends      'Test::Kwalitee'            => 0.30;
    
#no_index        'directory'         => 'demos';
#install_script  'bin/myscript';
    
auto_install;
WriteAll;
